### PR TITLE
EQL: transform query AST into queryDSL

### PR DESCRIPTION
**Pull request by costin**
Draft PR 

Generate the actual queryDSL but also keep track of generated fields(TBD whether that's needed or not).

### CODE REVIEW

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/planner/QueryFolder.java [hunk 1]
```diff
@@ -19,6 +26,60 @@ PhysicalPlan fold(PhysicalPlan plan) {
...
     @Override
     protected Iterable<RuleExecutor<PhysicalPlan>.Batch> batches() {
+        Batch fold = new Batch("Fold queries",
+                new FoldFilter()
+        );
+        Batch finish = new Batch("Finish query", Limiter.ONCE,
+                new PlanOutputToQueryRef()
+        );
+        
+        //return Arrays.asList(fold, finish);
```
Comment (rw-access) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/planner/QueryFolder.java]: I'm assuming this is intentional and will be uncommented out later?

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/planner/QueryFolder.java [hunk 2]
```diff
@@ -19,6 +26,60 @@ PhysicalPlan fold(PhysicalPlan plan) {
...
         return emptyList();
     }
-}
+    
+    private static class FoldFilter extends FoldingRule<FilterExec> {
+        @Override
+        protected PhysicalPlan rule(FilterExec plan) {
+
```
Comment (aleksmaus) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/planner/QueryFolder.java]: nit: empty line, ocd kicks in :-)

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/querydsl/container/QueryContainer.java [hunk 3]
```diff
@@ -5,9 +5,164 @@
...
+        if (fieldAttr.field().isAlias() == false) {
+            while (actualField.parent() != null
+                    && actualField.parent().field().getDataType() != DataTypes.OBJECT
+                    && actualField.parent().field().getDataType() != DataTypes.NESTED
+                    && actualField.field().getDataType().hasDocValues() == false) {
+                actualField = actualField.parent();
+            }
+        }
+        while (rootField.parent() != null) {
+            fullFieldName.insert(0, ".").insert(0, rootField.parent().field().getName());
```
Comment (aleksmaus) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/querydsl/container/QueryContainer.java]: assuming fullFieldName is going to be used when topHitFieldRef impl is complete

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/querydsl/container/QueryContainer.java [hunk 4]
```diff
@@ -5,9 +5,164 @@
...
+            return true;
+        }
+
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+
+        QueryContainer other = (QueryContainer) obj;
+        return Objects.equals(query, other.query);
+    }
```
Comment (aleksmaus) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/querydsl/container/QueryContainer.java]: so we care about the rest of members of QueryContainer for hashCode and equals?

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/querydsl/container/QueryContainer.java [hunk 5]
```diff
@@ -5,9 +5,164 @@
...
+
+    @Override
+    public String toString() {
+        try (XContentBuilder builder = JsonXContent.contentBuilder()) {
+            builder.humanReadable(true).prettyPrint();
+            SourceGenerator.sourceBuilder(this, null, null).toXContent(builder, ToXContent.EMPTY_PARAMS);
+            return Strings.toString(builder);
+        } catch (IOException e) {
+            throw new RuntimeException("error rendering", e);
```
Comment (aleksmaus) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/querydsl/container/QueryContainer.java]: maybe something derived from RuntimeException? or is it fine for a draft for now?

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/planner/QueryFolder.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/planner/QueryFolder.java]: Yes.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/querydsl/container/QueryContainer.java (hunk 4 above)
Comment (costin) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/querydsl/container/QueryContainer.java]: Fixed

**Review by astefan — APPROVED** [review]
LGTM

**Pull request merged**